OBS Browser Source Popout Tracker

MeemawHustlin opened the pull request:
In the settings page, I've added a Browser Source Tracker section where the user can set their desired width/height to match their exact stream layout dimensions and generate a popout window link for use as a browser source (NOT a window source as transparency is not retained) in OBS or other similar software.

Eventually improvements:

- Make the columns setting work.
- For people using a dual PC setup we need a way to inject the spectate data into the popout link optionally.
- Possibly make the whole thing flex css that changes automatically based on size, or resize text situationally if the user sets too small of a width/height.

The window generated is static, meaning that manually resizing it does not change its appearance at all. This is intentional so that when it is added to OBS, it is the exact size it needs to be already.

MeemawHustlin committed:
Basic popout page set up

- Added a new popout.html page which is a very simplified version of the checklist.html that uses its own popout.css styles.
- Added checklist.js code to make it behave differently for the popout page specifically.

MeemawHustlin committed:
updated popout layout

Now designed to work as an OBS browser source instead of chroma keyed.

- Background of page is transparent.
- Added a font from the game.
- Made the popout by two columns for now.

Additionally, made wayshrines worth 0 and moved the remaining points to the greater powers section since wayshrines are not tracked by save file drag/drop and are not in themselves specifically a requirement.

In the future, I plan on adding settings for the following:
- Width/height
- single column, double column

MeemawHustlin committed:
set up window params for modifying popout

- columns don't do anything yet. Not sure how to implement this currently.
- User settings do properly pass to window params and they are read properly, but certain settings seem to not work if they are too small of values. Need to investigate and possibly redo a lot of the css here to make things more automatically formatted.

MeemawHustlin committed:
Added copy to clipboard for popup window link

- Added a box that shows a browser source link to copy. Doubles as a backup in case popup windows are blocked by the user's browser. Ultimately they only need the link for use in OBS anyway.
- Simplified the other copy button for the share code because it was needlessly complicated.
- Disabled the column option until I figure out how to solve it.

MeemawHustlin committed:
Made progress bar red less harsh

MichaelEbert commented on checklist.js:
```diff
@@ -6,21 +6,23 @@ function init(){
-		}
-		catch{
-			debugger;
+		if(window.location.pathname.substring(window.location.pathname.lastIndexOf("/") + 1).split(".")[0] != "popout"){
```
mmm no

MeemawHustlin replied:
I'm not really sure how to do this differently, all I know is that the popout does work if I enclose this part in this if statement and it doesn't work at all if I don't. Is it the verbose conditional that is the problem or that the conditional interrupts the regular init()? It seems to have no negative effect on the other pages.

MichaelEbert replied:
I don't like changing functionality based on the window.location.pathname.
Can this go in a separate script? or common functionality extracted into another file?

MeemawHustlin replied:
The real issue is I don't actually know enough about how this part of the code works to pull specific things out lol. Currently the tracker doesn't update unless you refresh manually anyway and it automatically mirrors the spectate/local data, so if I understood the checklist and share code better I would have a "Local" and "Shared" tracker setting as a window param for the user based on if they have their stream PC separate from their checklisting PC, and somehow the local copy would have to know when to update. The resulting code for this would likely be how you are suggesting, with a separate script that borrows relevant code from the checklist.js and share.js files.

MichaelEbert replied:
if user is using the webpage outside of OBS, they'll need to spectate in the OBS browser to get the data.
If the user is using the webpage in browser source to check off items, then loadProgressFromCookie() (potentially followed by recalculateProgress()) should do the trick

MichaelEbert commented on popout.html:
```diff
@@ -0,0 +1,48 @@
+	<link rel="stylesheet" type="text/css" href="css/popout.css"/>
+	<script src="./js/module-shim.mjs" type="module"></script>
+	<script src="https://valarnin.github.io/oblivion-save-reader/dist/bundle.savefile.js" defer></script>
+    <script src="https://valarnin.github.io/oblivion-save-reader/dist/bundle.record.js" defer></script>
```
only need these if we're going to be parsing saves on this page

MeemawHustlin replied:
removed them, thanks for point that out

MichaelEbert commented on popout.js:
```diff
@@ -0,0 +1,35 @@
+//Runs in the settings.html page when the user clicks the "Generate Browser Source Tracker" button. Generates a popout window and passes params.
+function initBrowserSource(){
```
This should be a .mjs file

MeemawHustlin replied:
Anything I need to do besides rename the file? Should it go in that js folder with the other .mjs files or anything?

MichaelEbert replied:
yeah in the js/ folder. mjs are a little different. they load asynchronously. and need to export functions used by other scripts. and can only be called by other js module code.

MeemawHustlin replied:
Alright I've converted it and adjusted the settings/popout pages accordingly, thanks for explaining how it works

MichaelEbert commented on settings.html:
```diff
@@ -73,6 +74,26 @@ <h1>Settings</h1>
 </div>
 
+<div class="section">
+<div class="sectionTitle">Browser Source</div>
```
maybe "mini tracker"? calling this feature "browser source" is like calling the checklist "html page".

"Generate a condensed tracking page that can be used in stream overlays."

MeemawHustlin replied:
I called it a Browser Source because that's what it is called in OBS and figured a user looking for a stream overlay would be looking specifically for that term, but I see what you mean. It looks like similar features in things like StreamLabs are called Widgets. Is calling it a Widget fine? Mini tracker seems not accurate enough because this popout isn't really intended to be used outside of broadcast software.

MichaelEbert replied:
"(stream) widget" works great.

MeemawHustlin committed:
Fixed several requested changes

- popout.js is now an .mjs file and settings/popout.html were both updated accordingly.
- Removed the automatic popout on generate, seems intrusive and some browsers straight up block this.
- Removed valarnin's save reader bundles from the head of the popout since they do nothing.
- Renamed the browser source feature to the Stream Widget

MeemawHustlin committed:
Set up settings for eventual share code popout

MichaelEbert committed:
moved popout init to its own function

MeemawHustlin committed:
Preparing for merging latest master in

New spectate mode fixes need to be merged in.

MeemawHustlin committed:
Merge branch 'master' into obs-popout

MeemawHustlin commented:
Alright with the latest spectate fixes in the master, this now works as far as I can tell!
